### PR TITLE
DiT device to host optimization

### PR DESCRIPTION
**Pull request by sosborne-TT**
### Ticket
[34079](https://github.com/tenstorrent/tt-metal/issues/34079) (VAE optimization).

### Problem description
On BH GLX device to host can be done more efficiently by directly copying device tensors to the host and concatenating on host. In addition we can avoid using the default `to_torch` call which makes a copy of the data.

For now, on multi-host systems it will fall back to the original implementation. A better approach is probably to use the changes in this PR to copy the device tensors to each host, then each host transfers the partial results to the rank-0 host.

### What's changed

The new device to host function:
- Issues async DMA for all per-device shards concurrently.
- Synchronizes once.
- Uses ``to_torch_with_padded_shape`` which wraps the host buffer directly as a PyTorch tensor (zero-copy for ROW_MAJOR layout).
- Trims tile padding and concatenates on host.

For the baseline Wan model we would expect the device to host transfer to take `(3 * 81 * 720 * 1280 * 2) / 64e9 = 7ms` assuming 64GB/s PCIe bandwidth and bfloat16 data type. Since we have removed additional host copies the only additional time should be concatenating the result.

Performance for the baseline Wan model (with 2 inference steps):
```
================================================================================`
WAN PERFORMANCE RESULTS
================================================================================
Image Size: 1280x720
Inference Steps: 2
Num Frames: 81
DiT Configuration: sp=8, tp=4
Mesh Shape: MeshShape([4, 8])
Topology: Topology.Ring
--------------------------------------------------------------------------------
Text Encoding             | Mean:   0.0833s | Std:   0.0000s | Min:   0.0833s | Max:   0.0833s
Denoising                 | Mean:   7.1419s | Std:   0.0000s | Min:   7.1419s | Max:   7.1419s
VAE Decoding              | Mean:   2.4832s | Std:   0.0000s | Min:   2.4832s | Max:   2.4832s
Total Pipeline            | Mean:   9.7608s | Std:   0.0000s | Min:   9.7608s | Max:   9.7608s
--------------------------------------------------------------------------------
```

With the changes in this PR:
```
================================================================================
WAN PERFORMANCE RESULTS
================================================================================
Image Size: 1280x720
Inference Steps: 2
Num Frames: 81
DiT Configuration: sp=8, tp=4
Mesh Shape: MeshShape([4, 8])
Topology: Topology.Ring
--------------------------------------------------------------------------------
Text Encoding             | Mean:   0.0829s | Std:   0.0000s | Min:   0.0829s | Max:   0.0829s
Denoising                 | Mean:   7.1067s | Std:   0.0000s | Min:   7.1067s | Max:   7.1067s
VAE Decoding              | Mean:   1.8642s | Std:   0.0000s | Min:   1.8642s | Max:   1.8642s
Total Pipeline            | Mean:   9.1033s | Std:   0.0000s | Min:   9.1033s | Max:   9.1033s
--------------------------------------------------------------------------------
```

### Checklist

- [x] [T3K Wan demo](https://github.com/tenstorrent/tt-metal/actions/runs/24105052253) passes
- [x] [Galaxy Wan demo](https://github.com/tenstorrent/tt-metal/actions/runs/24204909011/job/70660180808) passes

### CODE REVIEW

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 1]
```diff
@@ -1026,14 +1026,22 @@ def __call__(
...
             concat_dims = [None, None]
             concat_dims[self.vae_parallel_config.height_parallel.mesh_axis] = 3
             concat_dims[self.vae_parallel_config.width_parallel.mesh_axis] = 4
-            video_torch = self.vae_ccl_manager.device_to_host(tt_video_BCTHW, concat_dims)
+            video_torch = fast_device_to_host(tt_video_BCTHW, self.mesh_device, concat_dims)
             video_torch = video_torch[:, :, :, :new_logical_h, :]
 
             if output_type == "np":
-                video = (video_torch * 0.5 + 0.5).clamp(0, 1).permute(0, 2, 3, 4, 1).float().numpy()
+                video = video_torch.permute(0, 2, 3, 4, 1).float().numpy()
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:1052]: Fixed

File: models/tt_dit/utils/tensor.py [hunk 2]
```diff
@@ -247,6 +247,70 @@ def local_device_to_torch(tt_tensor: ttnn.Tensor) -> torch.Tensor:
...
+    mesh_shape = tuple(mesh_device.shape)
+    device_tensors = ttnn.get_device_tensors(tt_tensor)
+
+    # Async DMA: issue all transfers then sync once
+    host_tensors = [dt.cpu(blocking=False) for dt in device_tensors]
+    ttnn.synchronize_device(mesh_device)
+
+    # Zero-copy to_torch (wraps existing host buffer, no memcpy)
+    with ThreadPoolExecutor(max_workers=len(host_tensors)) as pool:
+        shards = list(pool.map(ttnn.Tensor.to_torch_with_padded_shape, host_tensors))
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/utils/tensor.py]: Fixed

File: models/tt_dit/utils/tensor.py [hunk 3]
```diff
@@ -247,6 +247,70 @@ def local_device_to_torch(tt_tensor: ttnn.Tensor) -> torch.Tensor:
...
+    elif concat_dims[0] is not None:
+        return torch.cat(
+            [shards[r * mesh_shape[1]] for r in range(mesh_shape[0])],
+            dim=concat_dims[0],
+        )
+    elif concat_dims[1] is not None:
+        return torch.cat(
+            [shards[c] for c in range(mesh_shape[1])],
+            dim=concat_dims[1],
+        )
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/utils/tensor.py:395]: Fixed

File: models/tt_dit/utils/tensor.py [hunk 4]
```diff
@@ -247,6 +247,70 @@ def local_device_to_torch(tt_tensor: ttnn.Tensor) -> torch.Tensor:
...
+            rows.append(torch.cat(row_shards, dim=concat_dims[1]))
+        return torch.cat(rows, dim=concat_dims[0])
+    elif concat_dims[0] is not None:
+        return torch.cat(
+            [shards[r * mesh_shape[1]] for r in range(mesh_shape[0])],
+            dim=concat_dims[0],
+        )
+    elif concat_dims[1] is not None:
+        return torch.cat(
+            [shards[c] for c in range(mesh_shape[1])],
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/utils/tensor.py]: Fixed

File: models/tt_dit/utils/tensor.py [hunk 5]
```diff
@@ -247,6 +247,128 @@ def local_device_to_torch(tt_tensor: ttnn.Tensor) -> torch.Tensor:
...
+    from concurrent.futures import ThreadPoolExecutor
+
+    import torch
+
+    mesh_shape = tuple(mesh_device.shape)
+
+    # Get mesh coordinates before issuing DMA — topology is on the original
+    # device tensor and maps each shard index to its (row, col) mesh position.
+    mesh_coords = list(tt_tensor.tensor_topology().mesh_coords())
+    device_tensors = ttnn.get_device_tensors(tt_tensor)
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/utils/tensor.py:331]: Addressed.

File: models/tt_dit/utils/tensor.py [hunk 6]
```diff
@@ -247,6 +247,128 @@ def local_device_to_torch(tt_tensor: ttnn.Tensor) -> torch.Tensor:
...
+        shards = list(pool.map(_to_torch_zero_copy, host_tensors))
+
+    # Trim physical (tile-padded) shape to logical shape — view, no copy
+    logical_shape = list(host_tensors[0].shape)
+    shards = [s[tuple(slice(0, d) for d in logical_shape)] for s in shards]
+
+    # Build coord→shard mapping using explicit mesh coordinates rather than
+    # assuming get_device_tensors() returns shards in row-major order.
+    shards_by_coord = {(int(c[0]), int(c[1])): s for c, s in zip(mesh_coords, shards)}
+
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/utils/tensor.py:359]: Addressed.

File: models/tt_dit/utils/tensor.py [hunk 7]
```diff
@@ -247,6 +247,154 @@ def local_device_to_torch(tt_tensor: ttnn.Tensor) -> torch.Tensor:
...
+        )
+
+    # Validate: if a mesh axis is not gathered (concat_dims[axis] is None),
+    # the tensor must be replicated along that axis (size 1), otherwise we'd
+    # silently drop shards.
+    for axis in range(len(concat_dims)):
+        if concat_dims[axis] is None and mesh_shape[axis] > 1:
+            msg = (
+                f"concat_dims[{axis}] is None (no gather) but mesh_shape[{axis}]={mesh_shape[axis]} > 1. "
+                f"This would drop shards. Either gather along this axis or ensure the tensor is replicated."
```
[the remark this answers is not in the document]
Comment (sosborne-TT) [reply models/tt_dit/utils/tensor.py]: Fixed.

**Review by cglagovichTT — APPROVED** [review]
